openstack/identity/v3/extensions/ec2tokens/requests.go

```diff
@@ -0,0 +1,320 @@
+		signedHeaders,
+		bodyHash,
+	}, "\n")
+	hash := sha256.Sum256([]byte(canonicalRequest))
+
+	strToSign := strings.Join([]string{
+		EC2CredentialsAwsHmacV4,
+		date.Format(EC2CredentialsTimestampFormatV4),
+		scope,
+		hex.EncodeToString(hash[:]),
+	}, "\n")
+
+	key := EC2CredentialsBuildSignatureKeyV4(opts.Secret, date.Format(EC2CredentialsDateFormatV4), opts.Region, opts.Service)
+
+	return hex.EncodeToString(sumHMAC256(key, []byte(strToSign)))
+}
+
+// ToTokenV3ScopeMap is a dummy method to satisfy tokens.AuthOptionsBuilder interface
+func (opts *AuthOptions) ToTokenV3ScopeMap() (map[string]interface{}, error) {
+	return nil, nil
+}
+
+// CanReauth is a method method to satisfy tokens.AuthOptionsBuilder interface
+func (opts *AuthOptions) CanReauth() bool {
+	return opts.AllowReauth
+}
+
+// ToTokenV3CreateMap formats an AuthOptions into a create request.
+func (opts *AuthOptions) ToTokenV3CreateMap(map[string]interface{}) (map[string]interface{}, error) {
+	b, err := gophercloud.BuildRequestBody(opts, "credentials")
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.Signature != nil {
+		return b, nil
+	}
+
+	// calculate signature, when it is not set
+	c, _ := b["credentials"].(map[string]interface{})
+	h := interfaceToMap(c, "headers")
+	p := interfaceToMap(c, "params")
+
+	// detect and process a signature v2
+	if v, ok := p["SignatureVersion"]; ok && v == "2" {
+		if _, ok := c["body_hash"]; ok {
+			delete(c, "body_hash")
+		}
+		if _, ok := c["headers"]; ok {
+			delete(c, "headers")
+		}
+		if v, ok := p["SignatureMethod"]; ok {
+			// params is a map of strings
+			strToSign := EC2CredentialsBuildStringToSignV2(*opts)
+			switch v {
+			case EC2CredentialsHmacSha1V2:
+				// keystone uses this method only when HmacSHA256 is not available on the server side
+				// https://github.com/openstack/python-keystoneclient/blob/stable/train/keystoneclient/contrib/ec2/utils.py#L151..L156
+				c["signature"] = sumHMAC1([]byte(opts.Secret), []byte(strToSign))
+				return b, nil
+			case EC2CredentialsHmacSha256V2:
+				c["signature"] = sumHMAC256([]byte(opts.Secret), []byte(strToSign))
+				return b, nil
+			}
+			return nil, fmt.Errorf("unsupported signature method: %s", v)
+		}
+		return nil, fmt.Errorf("signature method must be provided")
+	} else if ok {
+		return nil, fmt.Errorf("unsupported signature version: %s", v)
+	}
+
+	// it is not a signature v2, but a signature v4
+	date := time.Now().UTC()
+	if opts.Timestamp != nil {
+		date = *opts.Timestamp
+	}
+	if v, _ := c["body_hash"]; v == nil {
+		// when body_hash is not set, generate a random one
+		c["body_hash"] = randomBodyHash()
+	}
+
+	signedHeaders, _ := h["X-Amz-SignedHeaders"]
+	c["signature"] = EC2CredentialsBuildSignatureV4(*opts, signedHeaders, date, c["body_hash"].(string))
+
+	h["X-Amz-Date"] = date.Format(EC2CredentialsTimestampFormatV4)
+	h["Authorization"] = fmt.Sprintf("%s Credential=%s/%s/%s/%s/%s, SignedHeaders=%s, Signature=%s",
```

Comment:
@jtopjian what if we move `Authorization` header generation into a func?

Reply:
If I'm understanding correctly, all this is doing is building a string, right? And a function for this would have to have several parameters passed into it. Unless I'm misunderstanding the idea you have, I think this is fine the way it is.

Reply:
```diff
diff --git a/openstack/identity/v3/extensions/ec2tokens/requests.go b/openstack/identity/v3/extensions/ec2tokens/requests.go
index 9be866d1..233f00a3 100644
--- a/openstack/identity/v3/extensions/ec2tokens/requests.go
+++ b/openstack/identity/v3/extensions/ec2tokens/requests.go
@@ -180,6 +180,20 @@ func EC2CredentialsBuildSignatureV4(opts AuthOptions, signedHeaders string, date
        return hex.EncodeToString(sumHMAC256(key, []byte(strToSign)))
 }
 
+// EC2CredentialsBuildAuthorizationHeaderV4 builds an AWS v4 Authorization
+// header based on auth parameters, date and signature
+func EC2CredentialsBuildAuthorizationHeaderV4(opts AuthOptions, signedHeaders string, signature string, date time.Time) string {
+       return fmt.Sprintf("%s Credential=%s/%s/%s/%s/%s, SignedHeaders=%s, Signature=%s",
+               EC2CredentialsAwsHmacV4,
+               opts.Access,
+               date.Format(EC2CredentialsDateFormatV4),
+               opts.Region,
+               opts.Service,
+               EC2CredentialsAwsRequestV4,
+               signedHeaders,
+               signature)
+}
+
 // ToTokenV3ScopeMap is a dummy method to satisfy tokens.AuthOptionsBuilder interface
 func (opts *AuthOptions) ToTokenV3ScopeMap() (map[string]interface{}, error) {
        return nil, nil
@@ -239,24 +253,15 @@ func (opts *AuthOptions) ToTokenV3CreateMap(map[string]interface{}) (map[string]
        if opts.Timestamp != nil {
                date = *opts.Timestamp
        }
+       h["X-Amz-Date"] = date.Format(EC2CredentialsTimestampFormatV4)
        if v, _ := c["body_hash"]; v == nil {
                // when body_hash is not set, generate a random one
                c["body_hash"] = randomBodyHash()
        }
-
        signedHeaders, _ := h["X-Amz-SignedHeaders"]
-       c["signature"] = EC2CredentialsBuildSignatureV4(*opts, signedHeaders, date, c["body_hash"].(string))
-
-       h["X-Amz-Date"] = date.Format(EC2CredentialsTimestampFormatV4)
-       h["Authorization"] = fmt.Sprintf("%s Credential=%s/%s/%s/%s/%s, SignedHeaders=%s, Signature=%s",
-               EC2CredentialsAwsHmacV4,
-               opts.Access,
-               date.Format(EC2CredentialsDateFormatV4),
-               opts.Region,
-               opts.Service,
-               EC2CredentialsAwsRequestV4,
-               signedHeaders,
-               c["signature"])
+       signature := EC2CredentialsBuildSignatureV4(*opts, signedHeaders, date, c["body_hash"].(string))
+       c["signature"] = signature
+       h["Authorization"] = EC2CredentialsBuildAuthorizationHeaderV4(*opts, signedHeaders, signature, date)
 
        return b, nil
 }
```

Reply:
Yeah, a single-line function feels unnecessary to me.